Implement relationship discovery for OGR layers, and show relationships in browser

nyalldawson opened the pull request:
(requires GDAL 3.6 or later!)

This PR implements two user-facing changes:

- it allows the Project Properties - Relations - Discover Relations feature to work for OGR layers which support relationships. Specifically, this includes GeoPackage (via the related tables extension) and ESRI FileGeodatabases
- it shows detailed relationship information for these data sources in the browser panel (see screenshot below)
 
![image](https://user-images.githubusercontent.com/1829991/181653957-c725fd4b-1615-417c-8459-c6aa78c8077d.png)


I've included a bunch of related cleanups to the relationship classes in QGIS. There's two main parts to this:

1. A general cleanup including moving common enums to qgis.h and promoting to enum classes
2. An expansion of the QgsWeakRelation class, so that it is able to losslessly represent the full details of relationships available from the GDAL API (which was designed to be a "superset" of relationship properties from common formats)

The docs give a good breakdown on why this was added to QgsWeakRelation:

```
 * In constrast to QgsRelation, QgsWeakRelation can be used to encapsulate
 * information about a relationship which does not currently exist in a QGIS project.
 * E.g. it can be used to represent a relationship which exists in a database
 * backend (but not within a QGIS project). Accordingly, some properties
 * available in QgsWeakRelation are included for informational purposes only,
 * and cannot be translated to QgsRelations or respected in QGIS relationships.
```

So the primary distinction here is that a QgsWeakRelation represents a "theoretical" relationship from a datasource, whereas QgsRelation is the (lossy) functional version of that relationship in a QGIS project.

nyalldawson committed:
[feature] Implement relationship discovery for OGR layers

(requires GDAL 3.6 or later)

This change allows the Project Properties - Relations - Discover
Relations feature to work for OGR layers which support relationships.
Specifically, this includes GeoPackage (via the related tables
extension) and ESRI FileGeodatabases

nyalldawson committed:
Tighten include in QgsRelation

nyalldawson committed:
Remove unnecessary forward declare

nyalldawson committed:
Move relationship enums to Qgis, promote to enum class

nyalldawson committed:
Minor fixups

nyalldawson committed:
Nicer strings

nyalldawson committed:
Tighten includes

nyalldawson committed:
Fix modifying loop detaches

nyalldawson committed:
Fix build on older GDAL

nyalldawson committed:
Expose QgsWeakRelation to PyQGIS

nyalldawson committed:
Add QgsAbstractDatabaseProviderConnection interface for retrieving relationships for a database

nyalldawson committed:
Implement relationship retrieval for ogr provider connection

nyalldawson committed:
Show existing relationships in browser when expanding datasources

nyalldawson committed:
Add additional properties to QgsWeakRelation to match those available
through GDAL api

nyalldawson committed:
Add some methods to get translated strings for relationship enums

nyalldawson committed:
Rework QgsWeakRelation so that it can losslessly represent many-to-many relations

nyalldawson committed:
Show detailed relationship metadata in browser dock widget

nyalldawson commented:
Note that a follow up PR will implement new logic which auto-adds related tables when a parent table is added from an OGR source (and automatically creates the corresponding relationship)

nyalldawson committed:
Fix build on older gdal

rouault commented on src/core/providers/ogr/qgsogrproviderconnection.cpp:
```diff
@@ -858,4 +863,177 @@ QgsAbstractDatabaseProviderConnection::SqlVectorLayerOptions QgsOgrProviderConne
+  }
+#else
+  Q_UNUSED( tableName )
+  throw QgsProviderConnectionException( QObject::tr( "Retrieving relationships for datasets requires GDAL 3.5 or later" ) );
```
```suggestion
  throw QgsProviderConnectionException( QObject::tr( "Retrieving relationships for datasets requires GDAL 3.6 or later" ) );
```

nyalldawson committed:
Also show a relationships item when expanding out tables for
tables which form the parent side of a relationship

nyalldawson committed:
Fix version warning

nyalldawson committed:
Rework QgsWeakRelation::resolvedRelation to instead return a list
of resolved relationships, so that it can handle resolving
a many-to-many weak relationship to two QgsRelations (which is
required to represent a many-to-many relationship using QgsRelation)

nirvn commented on src/app/qgsrelationadddlg.cpp:
```diff
@@ -141,17 +139,17 @@ void QgsRelationAddDlg::updateFieldsMappingHeaders()
 }
 
-QString QgsRelationAddDlg::referencingLayerId()
+QString QgsRelationAddDlg::referencingLayerId() const
```
This makes me think we should harmonize our use of Dlg vs Dialog class name in QGIS 4.0.

In the meantime, might be worth killing the _only_ use of dlg in src/app, that is QgsRelationAddDlg :) doesn't have to happen in this PR.

nyalldawson replied:
Ok, rename to QgsCreateRelationDialog

nirvn commented on src/core/browser/qgsrelationshipsitem.cpp:
```diff
@@ -0,0 +1,123 @@
+        {
+          try
+          {
+            QgsRelationshipItem *relationshipItem { new QgsRelationshipItem( this, relation ) };
```
```suggestion
            QgsRelationshipItem *relationshipItem = new QgsRelationshipItem( this, relation );
```

nirvn commented on src/core/providers/ogr/qgsgeopackageproviderconnection.cpp:
```diff
@@ -336,6 +336,10 @@ void QgsGeoPackageProviderConnection::setDefaultCapabilities()
   mCapabilities |= Capability::ListFieldDomains;
 #endif
 
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(3,6,0)
```
This might well kick me into building gdal locally instead of relying on the distro package :)

nirvn commented on src/core/qgis.h:
```diff
@@ -2072,6 +2072,48 @@ class CORE_EXPORT Qgis
+     * \since QGIS 3.28
+     */
+    enum class RelationshipType SIP_MONKEYPATCH_SCOPEENUM_UNNEST( QgsRelation, RelationType ) : int
+      {
```
Weird that the auto-formatter didn't catch this one.

```suggestion
{
```

nyalldawson replied:
Ah, it always does that when the monkey patch macro is in use. Not much we can do there...

nirvn commented on src/core/qgis.h:
```diff
@@ -2072,6 +2072,48 @@ class CORE_EXPORT Qgis
+     * \since QGIS 3.28
+     */
+    enum class RelationshipStrength SIP_MONKEYPATCH_SCOPEENUM_UNNEST( QgsRelation, RelationStrength ) : int
+      {
```
```suggestion
{
```

nirvn commented on src/core/qgsogrutils.cpp:
```diff
@@ -998,6 +998,8 @@ QgsFields QgsOgrUtils::stringToFields( const QString &string, QTextCodec *encodi
 QStringList QgsOgrUtils::cStringListToQStringList( char **stringList )
 {
   QStringList strings;
+  if ( !stringList )
```
What about moving that _above_ the QStringList strings to save a few CPU cycles? :)

nirvn commented on src/core/qgspolymorphicrelation_p.h:
```diff
@@ -30,11 +30,12 @@
-
 #include <QSharedData>
 #include <QPointer>
+#include "qgis.h"
```
Can we move QGIS headers to site above the Qt ones?

nirvn commented on src/core/qgsweakrelation.cpp:
```diff
@@ -17,64 +17,196 @@
+      relation.setName( mRelationName );
+      relation.setStrength( mStrength );
+      QgsVectorLayerRef referencedLayerRef { mReferencedLayer };
+      QgsMapLayer *referencedLayer { referencedLayerRef.resolveWeakly( project, matchType ) };
```
There's a bunch of these which if it was up to me I'd re-write as QgsMapLayer *l = ...; instead. Feel free to ignore.

nirvn commented on src/core/qgsweakrelation.cpp:
```diff
@@ -17,64 +17,196 @@
+{
+  if ( QgsProviderMetadata *metadata = QgsProviderRegistry::instance()->providerMetadata( mReferencedLayer.provider ) )
+  {
+    return metadata->decodeUri( mReferencedLayer.source ).value( QStringLiteral( "layerName" ) ).toString();
```
I love how we just keep using this decodeUri more, and more, and more :)

nirvn commented on tests/src/core/testqgsweakrelation.cpp:
```diff
@@ -66,72 +67,172 @@ void TestQgsWeakRelation::cleanup()
+                           QStringLiteral( "memory" )
                          );
-  QVERIFY( ! weakRel.resolvedRelation( QgsProject::instance(), QgsVectorLayerRef::MatchType::Name ).isValid() );
+  weakRel.setReferencingLayerFields( {"fk_province" } );
```
```suggestion
  weakRel.setReferencingLayerFields( { "fk_province" } );
```

nirvn commented on tests/src/core/testqgsweakrelation.cpp:
```diff
@@ -66,72 +67,172 @@ void TestQgsWeakRelation::cleanup()
                          );
-  QVERIFY( ! weakRel.resolvedRelation( QgsProject::instance(), QgsVectorLayerRef::MatchType::Name ).isValid() );
+  weakRel.setReferencingLayerFields( {"fk_province" } );
+  weakRel.setReferencedLayerFields( {"pk" } );
```
Auto-formatter should pick those up if possible.

```suggestion
  weakRel.setReferencedLayerFields( { "pk" } );
```

nirvn commented on tests/src/core/testqgsweakrelation.cpp:
```diff
@@ -66,72 +67,172 @@ void TestQgsWeakRelation::cleanup()
+                           QStringLiteral( "Polygon?crs=epsg:4326&field=pk:int&field=province:int&field=municipality:string" ),
+                           QStringLiteral( "memory" )
+                         );
+  weakRel.setReferencingLayerFields( {"fk_province" } );
```
```suggestion
  weakRel.setReferencingLayerFields( { "fk_province" } );
```

nirvn commented on tests/src/core/testqgsweakrelation.cpp:
```diff
@@ -66,72 +67,172 @@ void TestQgsWeakRelation::cleanup()
+                           QStringLiteral( "memory" )
+                         );
+  weakRel.setReferencingLayerFields( {"fk_province" } );
+  weakRel.setReferencedLayerFields( {"pk" } );
```
```suggestion
  weakRel.setReferencedLayerFields( { "pk" } );
```

nirvn reviewed (approved):
Lovely stuff.

nyalldawson committed:
Rename QgsRelationAddDlg to QgsCreateRelationDialog

nyalldawson committed:
Formatting fixes

nyalldawson commented:
Thanks @nirvn, all addressed!

phidrho commented:
> Note that a follow up PR will implement new logic which auto-adds related tables when a parent table is added from an OGR source (and automatically creates the corresponding relationship)

Hi @nyalldawson,
I just updated GDAL to 3.6.0, and noticed automatic creation of relationship and adding child tables. I have noticed some things related to this but I'm not sure if they are bugs or is this behavior by design:

![mtm example](https://user-images.githubusercontent.com/1849293/201688260-993c8b3f-cffc-4a08-80aa-d261bfc7662e.png)

1) I created an example with many-to-many relationship and it confused me why it doesn't work other way around - when adding child features why doesn't it automatically add parent features and sets relationship in project configuration (in case of many-to-many relationship).

2) Other thing I noticed that I get `WARNING` errors `Missing Relation in configuration` for mapping table when identifying features - either on child or parent feature.
![warning message on every identify event](https://user-images.githubusercontent.com/1849293/201688919-f54452db-b43f-4bd5-b95e-3b59c0fa9cdd.png)

3) In `Browser` there is a ***empty*** field listed on mapping table with `Type: geometry` that actually doesn't exist in GPKG
![empty layer](https://user-images.githubusercontent.com/1849293/201688438-ed1937fe-9fc5-4efb-adb2-ced3b44de6b7.png)

 I can open bug report with data and demo project if needed, but I wasn't sure if these are bugs but are by design (expected).

nyalldawson commented:
@phidrho looks like your gpkg relationship doesn't follow the gpkg related tables specifications correctly. Can you share so that i can investigate? How was the gpkg relationship created?

phidrho commented:
> @phidrho looks like your gpkg relationship doesn't follow the gpkg related tables specifications correctly. Can you share so that i can investigate? How was the gpkg relationship created?

@nyalldawson


here is the raw gpkg - created with QGIS - [raw_data.zip](https://github.com/qgis/QGIS/files/10010386/raw_data.zip),
and here are queries for adding relations:

# GPKG many-to-many relations test

## 1. I created some demo layers via QGIS GUI and stored them in GPKG "raw_data.gpkg"

- 2 related layers (both of type **features**)
  - my_points
  - my_polygons

![related polygons and points](https://user-images.githubusercontent.com/1849293/201860740-51cd0757-cd15-4c2a-89e8-20a1ef0a7e2a.png)

## 2. Manually added information about relations via SQL

- I followed [OGC GeoPackage Related Tables Extension](https://docs.ogc.org/is/18-000/18-000.html)

### 2.1 Added relation schema by defined in extension

#### 2.1.1 Defining extension

```sql
INSERT INTO gpkg_extensions (table_name, column_name, extension_name, definition, scope) VALUES
('relmap_points_polys', NULL, 'related_tables', 'TBD', 'read-write');
```

#### 2.1.2 Creating and populating relations table

```sql
CREATE TABLE 'gpkgext_relations' (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    base_table_name TEXT NOT NULL,
    base_primary_column TEXT NOT NULL DEFAULT 'id',
    related_table_name TEXT NOT NULL,
    related_primary_column TEXT NOT NULL DEFAULT 'id',
    relation_name TEXT NOT NULL,
    mapping_table_name TEXT NOT NULL UNIQUE
);
```

#### 2.1.2 Creating and populating relations data

I followed many-to-many example from [Annex B](https://docs.ogc.org/is/18-000/18-000.html#_example_informative):

```sql
CREATE TABLE relmap_points_polys (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    base_id INTEGER CONSTRAINT [fk_relmap_points_polys-my_polygons]
        REFERENCES my_polygons (fid)
        ON DELETE CASCADE
        ON UPDATE CASCADE
        DEFERRABLE INITIALLY DEFERRED,
    related_id INTEGER CONSTRAINT [fk_relmap_points_polys-my_points]
        REFERENCES my_points (fid)
        ON DELETE CASCADE
        ON UPDATE CASCADE
        DEFERRABLE INITIALLY DEFERRED,
    created_on INTEGER DEFAULT (strftime('%s'))
);


CREATE INDEX [idx_relmap_points_polys-base_id] ON relmap_points_polys (
    base_id
);


CREATE INDEX [idx_relmap_points_polys-related_id] ON relmap_points_polys (
    related_id
);


INSERT INTO gpkgext_relations (base_table_name, base_primary_column, related_table_name, related_primary_column, relation_name, mapping_table_name) VALUES
('my_polygons', 'fid', 'my_points', 'fid', 'features', 'relmap_points_polys');
```

Schema visualisation - as described in gpkgext_relations - same as in Annex B:

![schema](https://user-images.githubusercontent.com/1849293/201861012-db8633dd-b0fe-4ecf-9f06-a09b84926447.png)

### 2.2 Adding GPKG metadata about relation table

```sql
INSERT INTO
    gpkg_contents (
        table_name,
        data_type,
        identifier,
        description,
        last_change,
        srs_id
    )
VALUES
    (
        'relmap_points_polys',
        'attributes',
        'relmap_points_polys',
        'Relation table between points and polygons',
        strftime('%Y-%m-%dT%H:%M:%fZ', 'now'),
        '0'
    );


INSERT INTO
    gpkg_ogr_contents (table_name, feature_count)
VALUES
    ('relmap_points_polys', '0');


CREATE TRIGGER trigger_delete_feature_count_relmap_points_polys
    AFTER
        DELETE ON relmap_points_polys BEGIN
    UPDATE
        gpkg_ogr_contents
    SET
        feature_count = feature_count - 1
    WHERE
        lower(table_name) = lower('relmap_points_polys');
END;


CREATE TRIGGER trigger_insert_feature_count_relmap_points_polys
    AFTER
        INSERT ON relmap_points_polys BEGIN
    UPDATE
        gpkg_ogr_contents
    SET
        feature_count = feature_count + 1
    WHERE
        lower(table_name) = lower('relmap_points_polys');
END;
```

## 3. Adding some relational data

```sql
INSERT INTO relmap_points_polys (base_id,related_id) VALUES
(1,1),
(2,1),
(1,2),
(2,2),
(1,3),
(1,4),
(2,5),
(2,6);
```

nyalldawson commented:
@phidrho 

I'm not sure what I'm doing different, but the db created using your process works fine for me!

I've attached it here -- can you spot the difference?


[created.zip](https://github.com/qgis/QGIS/files/10099484/created.zip)

nyalldawson commented:
@phidrho 
> In Browser there is a empty field listed on mapping table with Type: geometry that actually doesn't exist in GPKG

This is fixed by https://github.com/qgis/QGIS/pull/51028 (it's harmless though -- it just shows at the moment for all non-spatial tables in gpkg)

phidrho commented:
> @phidrho
> 
> I'm not sure what I'm doing different, but the db created using your process works fine for me!
> 
> I've attached it here -- can you spot the difference?
> 
> [created.zip](https://github.com/qgis/QGIS/files/10099484/created.zip)

Hi @nyalldawson,

first of all, thank you for checking this.

Everything works for me also, it's nothing wrong with implementation in code but in logic. I think it should be upgraded to recognize relationship in both ways on `many-to-many` relationships.

There are two cases:
1) when I add new point I need to reference it to specific polygon/s
  - notes:
      - or none, if polygon/s currently doesn't exist - this means no data entry to `mapping table` in the moment of creating a point, but should be possible afterwards
      - this is OK because `points` and `polygons` tables don't reference mapping table by themself, there is only FK from `mapping table` to `points` and `polygons` table

2) when I add new polygon, I need to reference points that belong to it
  - same notes from above apply here

Currently software recognizes only one of these cases, depends how you define relation.

I'm demonstrating this by adding another relation to GPKG (this is obviously wrong approach, because we have two mapping tables that we need to keep in sync by ourselves, but just for demo it's OK), with additional mapping table:
[raw_data_both_direction.zip](https://github.com/qgis/QGIS/files/10101806/raw_data_both_direction.zip)
Now it works as expected, you can add `points` or `polygons` layer and relation is recognized.

So, long story short, when `many-to-many` relation is detected, it should automatically add both way relations to QGIS project, and when either of these two layers are added, other layer and mapping table should be also added to project.
It's in the middle of a feature request and bug report, certain logic is missing but nothing is wrong with it, that's why I posted it here in first place.

I can understand if currently there's no funding for it, we can always add this other direction manually in the meantime, but then we should open new feature request proposal for this scenario.

And we also still have these *WARNING messages*, at least on my side, not sure if they come from GDAL or QGIS - also reported here on PostGIS data - #48776

nyalldawson commented:
@phidrho 

> So, long story short, when many-to-many relation is detected, it should automatically add both way relations to QGIS project,

This should already be happening -- eg when I open my version of the raw_data my_polygons table, I get the following relationships auto created in the QGIS project:

![image](https://user-images.githubusercontent.com/1829991/204399290-e6328202-7162-4d2f-9eda-402f4872ab45.png)

Is this what you see?

I should point out that there ARE some rough edges with many-to-many relationships in QGIS, which were out of scope for this work. The goal here was simply to get the relationships setup automatically and unlock the existing QGIS relationship support (which is fantastic for 1 to many, slightly fragile for many-to-many!)

phidrho commented:
@nyalldawson 
> Is this what you see?

Yes it's same on my side, but only if I load `my_polygons` from GPKG, then both `mapping table` and `my_points` layers are also loaded and relations are set up as expected.

But if I load only `my_points` layer to empty project, none of these two relationships are defined, nor `my_polygons` and `mapping table` layers gets loaded automatically. It's a many-to-many relationship (which is correctly recognized in Browser panel) so I expected it to also work in this case.

With this example from my previous comment: [raw_data_both_direction.zip](https://github.com/qgis/QGIS/files/10101806/raw_data_both_direction.zip), it works either if you try to load only `my_points` or `my_polygons` to empty project - everything gets set up as expected (relations and other layers).
